tardis/montecarlo/src/integrator.c

```diff
@@ -213,62 +221,120 @@ _formal_integral(
-
+                  // TODO: e-scattering: in principle we also have to check
+                  // that dtau is <<1 (as assumed in Lucy 1999); if not, there
+                  // is the chance that I_nu_b becomes negative
                   for (;pline < storage->line_list_nu + size_line;
                        // We have to increment all pointers simultanously
                        ++pline,
                        ++pexp_tau,
-                       ++patt_S_ul)
+                       ++patt_S_ul,
+                       ++pJblue_lu)
                     {
                       if (*pline < nu_end)
+                      {
+                        // next resonance not in current shell
                         break;
-                      I_nu[p_idx] = I_nu[p_idx] * (*pexp_tau) + *patt_S_ul;
-
+                      }
+
+                      // Calculate e-scattering optical depth to next resonance point
+                      zend = storage->time_explosion / C_INV * (1. - *pline / nu);
+                      dtau += (zend - zstart) * escat_op;
+
+                      if (first == 1){
+                        // First contribution to integration
+                        // NOTE: this treatment of I_nu_b (given by boundary
+                        // conditions) is not in Lucy 1999; should be
+                        // re-examined carefully 
+                        I_nu_b[p_idx] = I_nu_b[p_idx] + dtau * (*pJblue_lu - I_nu_b[p_idx]);
+                        first = 0;
+                      }
+                      else{
+                        // Account for e-scattering, c.f. Eqs 27, 28 in Lucy 1999
+                        Jkkp = 0.5 * (*pJred_lu + *pJblue_lu);
+                        I_nu_b[p_idx] = I_nu_r[p_idx] + dtau * (Jkkp - I_nu_r[p_idx]);
+                        // this introduces the necessary offset of one element between pJblue_lu and pJred_lu
+                        pJred_lu += 1;
+                      }
+
+                      // Lucy 1999, Eq 26
+                      I_nu_r[p_idx] = I_nu_b[p_idx] * (*pexp_tau) + *patt_S_ul;
+
+                      // reset e-scattering opacity 
+                      dtau = 0;
+                      zstart = zend;
+                    }
+                    // Calculate e-scattering optical depth to grid cell boundary
+                    zend = storage->time_explosion / C_INV * (1. - nu_end / nu);
+                    dtau += (zend - zstart) * escat_op;
+                    zstart = zend;
+
+                    if (i < size_z-1){
+                      // advance pointers
+                      direction = shell_id[i+1] - shell_id[i];
+                      pexp_tau  += direction * size_line;
+                      patt_S_ul += direction * size_line;
+                      pJred_lu  += direction * size_line;
+                      pJblue_lu += direction * size_line;
                     }
                 }
-              I_nu[p_idx] *= p;
+              I_nu_r[p_idx] *= p;
```

Comment:
If nu is redder than the reddest line `I_nu_r[p_idx]` will always be `0`. So we have to account for the fact, that there may not be a line interaction. I propose to initialize `I_nu_r` as it physically does make sense in my opinion. `I_nu_r` is the intensity after a interaction. We assume a photosphere and emit radiation there. I think it makes sense to say "After they leave the photosphere there are at the red wing of some transition that happens at the boundary of the photosphere".
For the case that the ray does not intersect the photosphere, we initialize to 0. Here again I think we can apply the argument that we are at the red wing of a hypothetical transition after which our intensity is 0. Then we propagate through matter and apply the electron scattering.

Does this interpretation make sense? It would at least solve the problem and make the code a little bit easier (only Jkk has to be calculated in the first block)

What's left is to decide what happens in terms of electron scattering if there is no resonance.

Swapping the initialization to `I_nu_r` and changing the code in `first` does produce the correct spectrum for disabled electron scattering.